components/brave_wallet_ui/page/screens/swap/hooks/useSwap.ts

```diff
@@ -73,20 +84,43 @@ const hasDecimalsOverflow = (
   return decimalPlaces !== null && decimalPlaces > 0
 }
 
+const getTokenFromParam = (
+  contractOrSymbol: string,
+  network: BraveWallet.NetworkInfo,
+  tokenList: BraveWallet.BlockchainToken[]
+) => {
+  return tokenList.find(
+    (token) =>
+      (token.chainId === network.chainId &&
+        token.coin === network.coin &&
+        token.contractAddress.toLowerCase() ===
+          contractOrSymbol.toLowerCase()) ||
+      (token.chainId === network.chainId &&
+        token.coin === network.coin &&
+        token.contractAddress === '' &&
+        token.symbol.toLowerCase() === contractOrSymbol.toLowerCase())
+  )
+}
+
 export const useSwap = () => {
-  // State
-  const [fromAccount, setFromAccount] = useState<
-    BraveWallet.AccountInfo | undefined
-  >(undefined)
-  const [fromToken, setFromToken] = useState<
-    BraveWallet.BlockchainToken | undefined
-  >(undefined)
-  const [toAccountId, setToAccountId] = useState<
-    BraveWallet.AccountId | undefined
-  >(undefined)
-  const [toToken, setToToken] = useState<
-    BraveWallet.BlockchainToken | undefined
-  >(undefined)
+  // routing
+  const query = useQuery()
+  const history = useHistory()
+
+  // Queries
+  // FIXME(onyb): what happens when defaultFiatCurrency is empty
+  const { data: defaultFiatCurrency } = useGetDefaultFiatCurrencyQuery()
+  const { data: supportedNetworks } = useGetSwapSupportedNetworksQuery()
+  const { data: fullTokenList } = useGetCombinedTokensListQuery()
+  const { account: fromAccount } = useAccountFromAddressQuery(
+    query.get('fromAccountId') ?? undefined
+  )
+  // TODO: deprecate toAccountId in favour of toAddress + toCoin
+  const toAccountId = fromAccount?.accountId
+  const toCoinFromParams = query.get('toCoin') ?? undefined
+  const toCoin = toCoinFromParams ? Number(toCoinFromParams) : undefined
+  const toAddress = query.get('toAddress') ?? undefined
```

Comment:
nit: we don't need these until `toAccountId` is removed. You can access the address and coin from `toAccountId` already.